File: gareth-execution-ri/src/main/java/org/craftsmenlabs/gareth/execution/invoker/DefinitionFactory.java
package org.craftsmenlabs.gareth.execution.invoker;

import java.lang.reflect.Constructor;
import java.lang.reflect.InvocationTargetException;
import org.springframework.beans.factory.BeanExpressionException;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.context.ApplicationContext;
import org.springframework.stereotype.Service;

@Service
public class DefinitionFactory
{
	@Autowired
	ApplicationContext applicationContext;

	/**
	 * Create a instance for particular class (only zero argument constructors supported)
	 *
	 * @param clazz
	 * @return
	 * @throws IllegalAccessException
	 * @throws InvocationTargetException
	 * @throws InstantiationException
	 */
	public Object getInstanceForClass(final Class clazz)
		throws IllegalAccessException, InvocationTargetException, InstantiationException
	{

		Object definition = createDefinition(clazz);
		if (definition != null)
		{
			return definition;
		}

		Constructor constructor = null;
		Object declaringClassInstance = null;

		final boolean memberClass = clazz.isMemberClass();
		final int requiredConstructorArguments = memberClass ? 1 : 0; //

		if (memberClass)
		{
			declaringClassInstance = getInstanceForClass(clazz.getDeclaringClass());
		}
		for (final Constructor declaredConstructor : clazz.getDeclaredConstructors())
		{
			if (declaredConstructor.getGenericParameterTypes().length == requiredConstructorArguments)
			{
				constructor = declaredConstructor;
				break;
			}
		}
		// If a valid constructor is available
		if (null != constructor)
		{
			final Object instance;
			constructor.setAccessible(true);
			if (memberClass)
			{
				instance = constructor.newInstance(declaringClassInstance);
			}
			else
			{
				instance = constructor.newInstance();
			}
			return instance;
		}
		throw new InstantiationException(String.format("Class %s has no zero argument argument constructor", clazz));
	}

	private Object createDefinition(final Class clazz)
	{
		try
		{
			return applicationContext.getBean(clazz);
		}
		catch (BeanExpressionException e)
		{
			return null;
		}
	}

}

File: gareth-execution-ri/src/main/kotlin/org/craftsmenlabs/gareth/execution/services/definitions/DefinitionFactory.kt
package org.craftsmenlabs.gareth.execution.services.definitions

import org.springframework.beans.factory.BeanExpressionException
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.context.ApplicationContext
import org.springframework.stereotype.Service
import java.lang.reflect.Constructor

@Service
open class DefinitionFactory @Autowired constructor(val applicationContext: ApplicationContext) {

    fun getInstanceForClass(clazz: Class<*>): Any = getDefinitionFromContext(clazz) ?: createDefinitionByReflection(clazz)


    private fun createDefinitionByReflection(clazz: Class<*>): Any {
        var constructor: Constructor<*>? = null
        var declaringClassInstance: Any? = null

        val memberClass = clazz.isMemberClass
        val requiredConstructorArguments = if (memberClass) 1 else 0 //

        if (memberClass) {
            declaringClassInstance = getInstanceForClass(clazz.declaringClass)
        }
        for (declaredConstructor in clazz.declaredConstructors) {
            if (declaredConstructor.genericParameterTypes.size == requiredConstructorArguments) {
                constructor = declaredConstructor
                break
            }
        }
        // If a valid constructor is available
        if (null != constructor) {
            val instance: Any
            constructor.isAccessible = true
            if (memberClass) {
                instance = constructor.newInstance(declaringClassInstance)
            } else {
                instance = constructor.newInstance()
            }
            return instance
        }
        throw InstantiationException(String.format("Class %s has no zero argument argument constructor", clazz))
    }

    private fun getDefinitionFromContext(clazz: Class<*>): Any? {
        try {
            return applicationContext.getBean(clazz)
        } catch (e: BeanExpressionException) {
            return null
        }

    }

}